### ISSUE TITLE
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by paulshryock** [11ty/eleventy#531]
**TL;DR: Trying to generate 15000 HTML files crashes Node**

I'm using Eleventy to generate many HTML email files at once, which will be sent later.

`/src/emails/email.md` (this file generates all of the files because of `pagination`):

```
---
title: The Title
slug: the-title
pagination:
  data: recipients.people
  size: 1
  alias: recipient
permalink: ./{{ recipient.EmailAddress }}-{{ slug }}.html
layout: email
---

{{ recipient.EmailAddress }}
```

`/src/_data/recipients.json`:

```js
{
	"people": [
		{
			"User_Principal_Name": "Leslie Jones",
			"EmailAddress": "leslie.jones@mycompany.com",
			"Domain_File": "mycompany.com"
		},
		{
			"User_Principal_Name": "John Smith",
			"EmailAddress": "john.smith@company.com",
			"Domain_File": "company.com"
		},
		// several more records...
	]
}
```

When I run `npx eleventy`, I get the following files generated, which is the desired result:

- `/dist/leslie.jones@mycompany.com-the-title.html`
- `/dist/john.smith@company.com-the-title.html`
- ...

However, once I add more data to `/src/_data/recipients.json`--so much data that there's about 15000 records--when I run `npx eleventy`, Node crashes with the following error:

```bash
<--- Last few GCs --->

[67722:0x103800000]    13091 ms: Mark-sweep 1371.0 (1458.1) -> 1366.9 (1458.1) MB, 129.9 / 0.0 ms  (average mu = 0.228, current mu = 0.145) allocation failure scavenge might not succeed
[67722:0x103800000]    13265 ms: Mark-sweep 1371.9 (1458.1) -> 1368.9 (1458.1) MB, 159.0 / 0.0 ms  (average mu = 0.157, current mu = 0.085) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x21a9047cfb7d]
Security context: 0x273706b9d921 <JSObject>
    1: RegExp(aka RegExp) [0x273706b88089](this=0x273782f025b1 <undefined>,0x2737ed85b569 <Very long string[18076]>,0x27371b5859b9 <String[1]: i>)
    2: compile(aka compile) [0x2737028cda29] [/Users/pshryock/Desktop/InfoSec-Password-Violation/node_modules/linkify-it/index.js:~150] [pc=0x21a904b4dce6](this=0x273782f025b1 <undefined>,0x2737ed858fb1 <LinkifyIt...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x100063a65 node::Abort() [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 2: 0x100064104 node::errors::TryCatchScope::~TryCatchScope() [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 3: 0x10019d9a7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 4: 0x10019d944 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 5: 0x1005a2122 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 6: 0x1005a4653 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 7: 0x1005a0b88 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 8: 0x10059ed45 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
 9: 0x1005ab5ec v8::internal::Heap::AllocateRawWithLightRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
10: 0x1005ab66f v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
11: 0x10057d4f4 v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
12: 0x1006f55cb v8::internal::EscapeRegExpSource(v8::internal::Isolate*, v8::internal::Handle<v8::internal::String>) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
13: 0x1006f5277 v8::internal::JSRegExp::Initialize(v8::internal::Handle<v8::internal::JSRegExp>, v8::internal::Handle<v8::internal::String>, v8::base::Flags<v8::internal::JSRegExp::Flag, int>) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
14: 0x100865f51 v8::internal::Runtime_RegExpInitializeAndCompile(int, v8::internal::Object**, v8::internal::Isolate*) [/Users/pshryock/.nvm/versions/node/v11.6.0/bin/node]
15: 0x21a9047cfb7d
16: 0x21a90482499e
Abort trap: 6
```

Is there a way I can prevent Node from crashing, so it will generate the files?

Related to #513

**Comment by Ryuno-Ki** [comment]
@paulshryock As stop gap solution: https://github.com/angular/angular-cli/issues/13734#issuecomment-477455693

That is `node --max_old_space_size=8192 ngx eleventy` (maybe it works only with a locally installed `eleventy`, which needs to be referenced via `node_modules/.bin/eleventy`).

On what OS are you working here? The file path would indicate a macOS …

**Comment by Ryuno-Ki** [comment]
Or maybe trying to work with batch processing, i.e. not everything at once but chunking them somehow. Can't tell without more details about your template / layout files.

**Comment by paulshryock** [comment]
> @paulshryock As stop gap solution: [angular/angular-cli#13734 (comment)](https://github.com/angular/angular-cli/issues/13734#issuecomment-477455693)
> 
> That is `node --max_old_space_size=8192 ngx eleventy` (maybe it works only with a locally installed `eleventy`, which needs to be referenced via `node_modules/.bin/eleventy`).
> 
> On what OS are you working here? The file path would indicate a macOS …

@Ryuno-Ki Thanks for the responses! I tried both `node --max_old_space_size=8192 ngx eleventy` and `node --max_old_space_size=8192 npx eleventy` (**npx**, since I wasn't sure if **ngx** was a typo), which both resulted in:

```bash
internal/modules/cjs/loader.js:605
    throw err;
    ^

Error: Cannot find module '/Users/pshryock/Desktop/InfoSec-Password-Violation/ngx'
...
```

Then I tried `node --max_old_space_size=8192 node_modules/.bin/eleventy`, which took 39 seconds and then successfully finished without error, but created 0 html files: `Processed 0 files in 38.81 seconds`

I am running MacOS High Sierra (10.13.6).

> Or maybe trying to work with batch processing, i.e. not everything at once but chunking them somehow. Can't tell without more details about your template / layout files.

I think you're right that I would have to chunk them in smaller groups.

After speaking with my client about this, we've decided to take a different approach, so I no longer need to solve the problem.

Thanks again for your helpful suggestions!

**Comment by Ryuno-Ki** [comment]
Yeah, `npx`. Sorry for the typo. I'm still using `npm` for everything :-)
